File: go/datastructures/queue.go
package datastructures

// Queue implements a FIFO queue using a linked list.
// Time: O(1) for all operations
// Space: O(n)
type Queue[T any] struct {
	head *queueNode[T]
	tail *queueNode[T]
	size int
}

type queueNode[T any] struct {
	value T
	next  *queueNode[T]
}

func NewQueue[T any]() *Queue[T] {
	return &Queue[T]{}
}

func (q *Queue[T]) Enqueue(value T) {
	node := &queueNode[T]{value: value}
	if q.tail != nil {
		q.tail.next = node
	} else {
		q.head = node
	}
	q.tail = node
	q.size++
}

func (q *Queue[T]) Dequeue() (T, bool) {
	var zero T
	if q.head == nil {
		return zero, false
	}
	value := q.head.value
	q.head = q.head.next
	if q.head == nil {
		q.tail = nil
	}
	q.size--
	return value, true
}

func (q *Queue[T]) Peek() (T, bool) {
	var zero T
	if q.head == nil {
		return zero, false
	}
	return q.head.value, true
}

func (q *Queue[T]) IsEmpty() bool {
	return q.size == 0
}

func (q *Queue[T]) Size() int {
	return q.size
}

func (q *Queue[T]) Clear() {
	q.head = nil
	q.tail = nil
	q.size = 0
}

File: haskell/src/DataStructures/Queue.hs
{-|
Module      : DataStructures.Queue
Description : FIFO queue implementation using two stacks
Stability   : stable

A purely functional queue using the banker's queue technique.
Time: O(1) amortized for all operations
Space: O(n)
-}
module DataStructures.Queue
  ( Queue
  , empty
  , isEmpty
  , enqueue
  , dequeue
  , peek
  , size
  , fromList
  , toList
  ) where

-- | A queue using two lists for amortized O(1) operations
data Queue a = Queue [a] [a]
  deriving (Show, Eq)

-- | Create an empty queue
empty :: Queue a
empty = Queue [] []

-- | Check if queue is empty
isEmpty :: Queue a -> Bool
isEmpty (Queue [] []) = True
isEmpty _             = False

-- | Add an element to the back of the queue
enqueue :: a -> Queue a -> Queue a
enqueue x (Queue front back) = Queue front (x : back)

-- | Remove an element from the front of the queue
dequeue :: Queue a -> Maybe (a, Queue a)
dequeue (Queue [] [])     = Nothing
dequeue (Queue [] back)   = dequeue (Queue (reverse back) [])
dequeue (Queue (x:xs) back) = Just (x, Queue xs back)

-- | Peek at the front element without removing
peek :: Queue a -> Maybe a
peek (Queue [] [])   = Nothing
peek (Queue [] back) = Just (last back)
peek (Queue (x:_) _) = Just x

-- | Get the number of elements in the queue
size :: Queue a -> Int
size (Queue front back) = length front + length back

-- | Create a queue from a list (first element at front)
fromList :: [a] -> Queue a
fromList xs = Queue xs []

-- | Convert queue to list (front element first)
toList :: Queue a -> [a]
toList (Queue front back) = front ++ reverse back
